staging/src/k8s.io/legacy-cloud-providers/azure/azure.go

```diff
@@ -242,21 +273,65 @@ func NewCloud(configReader io.Reader) (cloudprovider.Interface, error) {
 		config.VMType = vmTypeStandard
 	}
 
+	if config.OverrideType == "" {
+		// The default override type is secretOverrideTypeCan.
+		config.OverrideType = secretOverrideTypeCan
+	} else {
+		supportedOverrideTypes := sets.NewString(
+			string(secretOverrideTypeCan),
+			string(secretOverrideTypeMust),
+			string(secretOverrideTypeNo))
+		if !supportedOverrideTypes.Has(string(config.OverrideType)) {
+			return fmt.Errorf("overrideType %v is not supported, supported values are %v", config.OverrideType, supportedOverrideTypes.List())
+		}
+	}
+
+	if config.ConfigType == "" {
+		// The default config type is secretConfigureAll.
+		config.ConfigType = secretConfigureAll
+	} else {
+		supportedConfigTypes := sets.NewString(
+			string(secretConfigureAll),
+			string(secretConfigureNode),
+			string(secretConfigureControlPlane))
+		if !supportedConfigTypes.Has(string(config.ConfigType)) {
+			return fmt.Errorf("configType %v is not supported, supported values are %v", config.ConfigType, supportedConfigTypes.List())
+		}
+	}
+
 	env, err := auth.ParseAzureEnvironment(config.Cloud)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	servicePrincipalToken, err := auth.GetServicePrincipalToken(&config.AzureAuthConfig, env)
 	if err == auth.ErrorNoAuth {
-		if !config.UseInstanceMetadata {
-			// No credentials provided, useInstanceMetadata should be enabled.
-			return nil, fmt.Errorf("useInstanceMetadata must be enabled without Azure credentials")
+		runingAsKubelet, err := isRunningAsKubelet()
+		if err != nil {
+			return err
+		}
+
+		// Controller manager could be initialized from secret.
+		if !runingAsKubelet && !fromSecret && az.Config.OverrideType != secretOverrideTypeNo {
+			klog.V(2).Infof("No credentials provided, would initialize from secret %s", getConfigSecretName(az.Config.ConfigType))
+			return nil
+		}
+
+		// No credentials provided, useInstanceMetadata should be enabled for Kubelet.
+		if runingAsKubelet && !config.UseInstanceMetadata {
+			return fmt.Errorf("useInstanceMetadata must be enabled without Azure credentials")
+		}
+
+		// Credentials are required for controller-manager.
+		if !runingAsKubelet && fromSecret {
```

Comment:
there are 3 conditions, can we make the logic a little more clear here?
```
if !runingAsKubelet && fromSecret
if !runingAsKubelet && !fromSecret && az.Config.OverrideType != secretOverrideTypeNo
if runingAsKubelet && !config.UseInstanceMetadata
```

e.g.
```
if runingAsKubelet {
 ...
} else {
 ...
}

```

Reply:
there're actually four conditions here: runingAsKubelet, fromSecret, UseInstanceMetadata and OverrideType. In that way, there would be a lot embeded if-conditions. Let me think how to make this more easier to read.

Reply:
Is this easier to read?
```
if runingAsKubelet {
    if !config.UseInstanceMetadata { ...}
} else {
   if fromSecret {
   ...
   } else {
      if az.Config.OverrideType != secretOverrideTypeNo { ...}
  }
}
```

Reply:
simplified the logic, PTAL